lib/liquid/context.rb

```diff
@@ -111,31 +107,15 @@ def pop
     #   end
     #
     #   context['var]  #=> nil
-    def stack(new_scope = nil)
```

Comment:
@this_stack_used seems like an old performance optimisation, however every where that stack is created there is a  very low chance it won't be used. Because of this read performance is taking a bigger hit than any cases where a scope may not be needed

Reply:
Plus this is adding 10+ lines of code and complexity which doesn't seem to be justified. Also makes reading and understanding this more complex than it needs to be.

Reply:
> this_stack_used seems like an old performance optimisation

https://github.com/Shopify/liquid/pull/400 has some discussion

Reply:
> every where that stack is created there is a very low chance it won't be used

What makes you say this? What proof do you have for this chance being "very low"? The original patch (https://github.com/Shopify/liquid/pull/400) was based on the assumption that this chance is high enough for this optimization to be worth it. I wonder what makes you think that that isn't the case.

Reply:
Interesting, I'm seeing the reverse result on benchmarks now by pulling this out.

Reply:
So looking in to this more I think this is solved on the wrong level. So each tag seems to be running `context.stack do` however there is no need to, any tag that doesn't set scope shouldn't even be running this. `context.stack do` should only be used by `for` etc where a new scope is needed. The only things that can change scope that are nested are `assign` and `capture` and both of these always write to the root level anyway

Reply:
So ultimately I think the right fix for this is to remove this but at the same time remove all unnecessary uses of `context.stack do`

Reply:
Then you should have 0 instances where you create a new scope and don't use it